### ISSUE TITLE
Allow publisher Nuant to use the Jupyter kernel API

### ISSUE THREAD

**Issue by aminelajmi** [microsoft/vscode-jupyter#13561]
Hi,

Following #13548, we would like to use the [vscode-jupyter-powertoys](https://github.com/microsoft/vscode-jupyter-powertoys) package to be able to access the default ipykernel programmatically from a custom notebook. Our extension id is **nuant.nuant-vscode-extension**

Thank you for the support.

**Comment by aminelajmi** [comment]
Hi, 

Is there an ETA for such a process? For now, we still fail to get the service that gives access to kernels:
```
        const kernelApi = (await api.getKernelService())!; // <--- returns undefined
```
Thanks for your time.

**Comment by DonJayamanne** [comment]
@aminelajmi 

Please understand that this API is still in a proposed state, meaning the API can change significantly, resulting in potential breakages of the extesnion using this API (such as yours).
The intention is to never break any extension using this API. The plan is to give consumers of the API some time to update their extension should there ever be a change to the proposed API in the Jupyter extension.

Please could you acknowledge you have understood this, thank you.

**Comment by DonJayamanne** [comment]
thanks, I'm assuming the name of the publisher is `nuant`, its the name of the publisher that is required.
Also, what methods in the API do you plan on using?

**Comment by aminelajmi** [comment]
Ideally, we'd like to submit python code and execute it using `IKernelConnection.requestExecute` (`requestDebug` is also nice to have). At runtime, we access `NotebookCell.outputs` to display the results.

Also, we'd like to share variables between Python and DSL, without extra "magic" commands, so intuitively, I'd need to access the notebook context programmatically (just like the variable explorer does) to be able to create/update variables. TBH, I don't know if this is allowed by the API, I'm still investigating.

**Comment by DonJayamanne** [comment]
Thank you for your patience, I have made a change so that anyone can review the API in VS Code Insiders.
https://github.com/microsoft/vscode-jupyter/pull/13663 (hopefully this will get merged some time today and a new version of Jupyter pre-release will be shipped tomorrow)

In the mean time you can install this VSIX into your VS Code insiders and try this out.
[ms-toolsai-jupyter-insiders.vsix (1).zip](https://github.com/microsoft/vscode-jupyter/files/11671251/ms-toolsai-jupyter-insiders.vsix.1.zip)

Basically you should be able to use the API in VS Code insiders without any problems. We do however display an error message, only to ensure users of the API reach out.

Once you have reviewed the API you can ping here and I will discuss with the team and we can add `nuant` into the allow list.
Apologies for this process, as mentioned its still proposed and we do not want to expose something that could change and break other extensions.

**Comment by aminelajmi** [comment]
Calling the API in VSCode Insiders together with the vsix extension still returns `undefined`. I can confirm however the error “Please contact the Jupyter Extension to get access to the Kernel API. Publisher nuant” is popping up… I suppose `nuant` is still not registered at this time.

**Comment by DonJayamanne** [comment]
@aminelajmi please can you try VS Code insiders with the latest pre-release version of the exetnsion.
Ensure you use vscode insiders and pre-release.
if you still run into issues, please could you share a screenshot of the error message

**Comment by aminelajmi** [comment]
Wonderful, it does return something now, I will continue to analyze the API and give you feedback soon.
Thanks a lot!

![Capture d’écran 2023-06-13 à 15 54 53](https://github.com/microsoft/vscode-jupyter/assets/81257433/d47ccb93-742b-45d7-b372-202c297050ab)

**Comment by DonJayamanne** [comment]
the types for the exposed API can be found here https://github.com/microsoft/vscode-jupyter/blob/main/src/api.d.ts

**Comment by DonJayamanne** [comment]
Once you have reviewed it, please do let me know what you plan on using and if you run into any issues,
then lets look into onborading your publisher to use this API in VS Code stable as well.

**Comment by aminelajmi** [comment]
Hello @DonJayamanne,

We are now able to obtain a kernel from kernel spec, and are able to submit code programmatically. 

However, we have encountered a challenge related to accessing the notebook variables. Upon inspecting the code base, we discovered that this particular functionality is implemented in Python:

```
// VariableTypesFunc takes in list of vars and the corresponding var names
const { code, cleanupCode, initializeCode } = await this.varScriptGenerator.generateCodeToGetVariableTypes({
    isDebugging: false
});
```

Based on our observations so far, it appears that the API does not provide a direct method to retrieve variables or -at least-give access to the `pythonVariableRequester` that accomplishes this task. 

Could you please suggest an alternative approach to retrieve the variable names and types?

Thank you for your valuable assistance.

**Comment by DonJayamanne** [comment]
> of course, we could do the same thing ourselves as we can submit arbitrary python code programmatically, but it's rather to

thanks for the feedback

unfortunately (at this stage) we do not plan on exposing such an api
Please feel free to copy the same code
I know that’s not the best, however for now let’s keep things simple (in the api)
Sort that doesn’t help you

**Comment by aminelajmi** [comment]
Hi @DonJayamanne, 

We are almost done with the review. We are able to retrieve the variables now by submitting python code to the kernel. Our plan is to industrialise this PoC and provide it as soon as possible to our early adopters.

Here are the API methods we are using so far:
```
- JupyterAPI.getKernelService
- IKernelConnection.requestExecute
- IKernelConnection.requestDebug
- IExportedKernelService.getKernelSpecifications
- IExportedKernelService.getActiveKernels
- IExportedKernelService.getKernel
- IExportedKernelService.startKernel
- IFuture<REQUEST extends KernelMessage.IShellControlMessage, REPLY extends KernelMessage.IShellControlMessage>.onReply
- IFuture<REQUEST extends KernelMessage.IShellControlMessage, REPLY extends KernelMessage.IShellControlMessage>.onIOPub
```

We would like to go further with the process of onboarding Nuant to be able to use the API in regular VS Code.

What would be the next move?

Thank you for the support

**Comment by DonJayamanne** [comment]
Thank you for getting back to us, I will add nuant into the list of alllowed publishers asap.
Unfortunately we're in the end game for our release (which will be next week), hence this change cannot make it into the next release (next week)
However it will be in the release after that.

**Comment by aminelajmi** [comment]
Hi @DonJayamanne,

We've have transitioned our IDE to be based on VSCode 1.81. Despite the update, we're still encountering restrictions when trying to access the Kernel API. Could you kindly clarify whether this situation is still being developed or if there's a step we might have overlooked?
<img width="990" alt="Capture d’écran 2023-08-10 à 22 43 44" src="https://github.com/microsoft/vscode-jupyter/assets/81257433/16da6879-d0da-448a-9e28-e4b93aca48cc">

Thank you for the support

**Comment by DonJayamanne** [comment]
What version of Jupyter extension do you have installed?

**Comment by DonJayamanne** [comment]
If possible, please could you share a copy of the VSIX and instructions to access features in the extension that would go down this path. Please email the VSIX (no need of the source, just the VSIX that you would otherwise publish to the marketplace) at the following = <don>.<jayamanne> then at microsoft.com.
Sending me the vsix should be safe as it would be the exact same content that one can download from the marketplace.

**Comment by aminelajmi** [comment]
Hi @Donjayamanne,

You are right: the version we are testing against is not the most up-to-date one. The reason behind this is that we are not making use of the official Microsoft marketplace; instead, we rely on Open VSX. To fully integrate it into our IDE, we need to grab version `2023.07` from the Open VSX marketplace.

Allow me to provide a bit more context for better understanding. Our IDE is built upon [OpenVSCode](https://github.com/gitpod-io/openvscode-server), and as a result, we retrieve extensions from https://open-vsx.org rather than the official marketplace. At present, the latest available Jupyter version on Open VSX is `2023.04` -> https://open-vsx.org/extension/ms-toolsai/jupyter

Could I kindly request that you update the Jupyter version on Open VSX for our usage?

I apologize for not clarifying this earlier. Thank you very much for your assistance!

**Comment by DonJayamanne** [comment]
> Could I kindly request that you update the Jupyter version on Open VSX for our usage?

unfortunately we do not publish extensions to the open vsix marketplace,
I don’t even how that works or who publishes extensions there

closing this for now

**Comment by DonJayamanne** [comment]
And thanks for the clarification
